### DISCUSSION TITLE
Improvements for linting with nx run-many and outputFile

### DISCUSSION THREAD

**Discussion by dennisluitwieler** [nrwl/nx#8769] [Feature Requests]
I have a suggestion for an improvement on using nx run-many for linting.

When you run lint with an outputFile, every module keeps overwriting the same outputfile, so you end up with the result of the last library being processed. For example:

`nx run-many --all --target=lint --outputFile=./output.json --format=json`

**Possible improvements:**

- Option 1: Running the linter on multiple projects combines the results of each module into 1 output file (in this case in json format). This would require no change to the command itself, but only to the actual outputFile.

- Option 2: The run-many command supports some wildcard or regex option to change the outputFile for each library (for instance having the library name as part of the output filename. This would make it possible for other tooling to combine these files into 1 combined json file. This would require some token or wildcard kind-of support for the outputFile parameter.

_Examples:_
```
nx run-many --all --target=lint --outputFile=./report-{$LIBRARY_NAME}.json --format=json
nx run-many --all --target=lint --outputFile=./reports/{project.name}/report.json --format=json

```

- Option 3: The run-many command supports passing an outputDir instead outputFile, which automatically generates multiple outputfiles in 1 dir.

_Example:_
`nx run-many --all --target=lint --outputDir=./report/ --format=json`

- Option 4: A combination of the above, perhaps making it configurable how you want your output files to be.

_Example:_

```
# Generates multiple files named after the library
nx run-many --all --target=lint --outputFilePattern=./reports/${project.name}.json --format=json
nx run-many --all --target=lint --outputDir=./reports --useProjectAsOutputFilename --format=json

# Combines all reports into 1 file
nx run-many --all --target=lint --outputFile=./report.json --format=json
nx run-many --all --target=lint --outputFile=./report.json --combineReports --format=json
```

**Use case**
The output of the linter can be used to feed a complete linter report to static code analysis tools such as SonarQube.
I believe this is a very common usecase that developers keep having to build themselves. At least for us, it is definitely a much needed feature, but our only option at the moment is to create a custom command that dynamically calls the linter for each module with a different output filename and then use some tooling to merge the reports.

It would be awesome if it would be a feature supported by nx.

**Comment by kauppfbi** [comment]
As far as I understand your wish, I would say, that this is already possible.

The builder `@nrwl/linter:eslint` already supports outputFiles: 
![image](https://user-images.githubusercontent.com/14169258/152039716-403b899c-32e0-46ea-9c87-6a4a837c89f0.png)

The trick is to configure it before you run `nx run-many ...`. And if you do not always want to have the outputFile or report format, you can simply create a dedicated configuration, e.g. for the ci. 

You can then run something like: `nx run-may --target=lint --configuration=ci`. Tools like SonarQube can them import the reports with a glob pattern.

**Reply by dennisluitwieler** [reply]
Wow, not sure how I overlooked that this is the way to make it possible, but I got it working. I guess this means  feature request can be marked as 'works for me'.

Thank you @kauppfbi ...

**Reply by thomas-lucas1985** [reply]
Hi. Just stumbled on this issue as well. If I understand correct, I need to set this configuration in each project.json for each lib I have? I understand this might work if you only have a few libs, but we're already on 40+ and still growing. Being able to configure/set this in 1 default place would be more convenient. Or do I miss something?

Thanks

**Reply by dennisluitwieler** [reply]
We manually changed our existing project.json files (Regex search & replace in your IDE might help). Furthermore, we also have a custom library generator that extends an existing generator and adds specific configuration for us, like library tags and also this extra lint configuration. We first created a completely custom target, but recently we removed that and added a configurations entry under the default lint entries so we can do something like

nx run affected:lint —configurations=ci

**Reply by thomas-lucas1985** [reply]
Thanks for your feedback. The issue was indeed not updating the existing libs, but to have a consistent system in place. I'll need to look at the generators. I haven't done any customisation before. I was also looking to put a configuration in general via the targetDefaults in nx.json, but they don't accept a configuration object. At least not in the version we're on (14.6).

**Reply by dennisluitwieler** [reply]
What we have is not ideal yet though. We had to set a custom output filename for the report inside each project.json file or else that same file gets overwritten by every library, would be simpler if we could for instance just refer to the current project target and use that as (part of) the filename. 

Something similar to `{options.outputFile}` like `{project.target}`

So we could do

"outputFile": ./reports/{project.target}.json"

**Reply by thomas-lucas1985** [reply]
Yeah, this is also what I fear and don't like. In the nx.json for the targetDefaults, they've foreseen variables like workspaceRoot,projectRoot and projectName. But unfortunately, this type of configuration doesn't seem to be allowed on this level.

**Comment by rujorgensen** [comment]
Has anyone found a solution that avoids editing each configuration file individually, or did NX make this possible since this discussion was opened 🙂? We're currently managing around 250 libraries.

**Reply by kennethvc** [reply]
I've tried it just now and I was able to use these placeholders. The following command results in a file per lib:

`nx run-many --all --target=lint --outputFile {projectName}.json`

**Reply by rujorgensen** [reply]
Excellent! I wasn't aware of the syntax involving `{projectName}`, it works great. In the end this was exactly what I needed: `nx run-many --all --target=lint --outputFile={projectName}.json --format=json`. Thanks a lot 🙏!
